example/imaginary-wasm.hh

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <functional>
+#include <initializer_list>
+#include <string>
+#include <vector>
+
+namespace wasm {
+struct Frame {
+  // TODO: what are the pointer invalidation rules here?
+  const Instance *instance() const;
+
+  size_t module_offset() const;
+  uint32_t func_index() const;
+  size_t func_offset() const;
+};
+
+struct Error : std::runtime_error {};
+
+// An error that occurs with a wasm function in the call stack.
+struct Trap : Error {
+  typedef iterator Frame*;
+  typedef const_iterator const Frame*;
+  Frame origin();
+  iterator begin();
+  iterator end();
+  const_iterator begin() const;
+  const_iterator end() const;
+};
+struct EngineConfig {
+  EngineConfig();
+  // Implementation provides extra c'tors here.
+};
+struct Module;
+struct Instance;
+struct Imports;
+struct Global;
+struct Store {
+  explicit Store();
+  explicit Store(const EngineConfig &);
+
+  Module compile(const std::vector<char> &);
+  Instance instantiate(const std::vector<char> &);
```

Comment:
You know better than I do, but is `char` the normal way that vectors of bytes are expressed in C++?  I know it's correct, just surprised there's not a fancier new type that is commonly used

Reply:
It is! There is also a newer fancier type named "std::byte" which is absolutely not the right thing to use here. It's safe to say that files in the C++ sense are contiguous chars (or unsigned chars).